frontend/src/api/APIClient.ts

```diff
@@ -0,0 +1,119 @@
+/* eslint-disable class-methods-use-this */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import ApiError from './ApiError';
+
+type Method = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
+
+type BodyHashMap = Record<string, any>;
+
+interface BaseAPIClientParams {
+  path: string;
+  hasCookies?: boolean;
+}
+
+interface APIClientParamsWithBody extends BaseAPIClientParams {
+  body?: BodyHashMap;
+}
+
+interface APIClientType {
+  get<T>(params: BaseAPIClientParams): Promise<T>;
+  post<T>(params: APIClientParamsWithBody): Promise<T>;
+  patch<T>(params: APIClientParamsWithBody): Promise<T>;
+  put<T>(params: APIClientParamsWithBody): Promise<T>;
+  delete(params: BaseAPIClientParams): Promise<void>;
+}
+
+export default class APIClient implements APIClientType {
+  private baseURL: URL;
+
+  constructor(baseURL: string) {
+    this.baseURL = new URL(baseURL);
+  }
+
+  async get<T>(params: BaseAPIClientParams): Promise<T> {
+    return this.request<T>({ method: 'GET', ...params });
+  }
+
+  async post<T>(params: APIClientParamsWithBody): Promise<T> {
+    return this.request<T>({ method: 'POST', ...params });
+  }
+
+  async patch<T>(params: APIClientParamsWithBody): Promise<T> {
+    return this.request<T>({ method: 'PATCH', ...params });
+  }
+
+  async put<T>(params: APIClientParamsWithBody): Promise<T> {
+    return this.request<T>({ method: 'PUT', ...params });
+  }
+
+  async delete(params: BaseAPIClientParams): Promise<void> {
+    return this.request<void>({ method: 'DELETE', ...params });
+  }
+
+  private getRequestInit({
+    method,
+    body,
+    hasCookies = true,
+  }: {
+    method: Method;
+    body?: BodyHashMap;
+    hasCookies?: boolean;
+  }) {
+    const headers: HeadersInit = {
+      Accept: 'application/json',
+    };
+
+    const requestInit: RequestInit = {
+      method,
+      credentials: hasCookies ? 'include' : 'omit',
+      headers,
+    };
+
+    if (['POST', 'PUT', 'PATCH'].includes(method)) {
+      headers['Content-Type'] = 'application/json';
+    }
+
+    if (body) {
+      requestInit.body = JSON.stringify(body);
+    }
+
+    return requestInit;
+  }
+
+  private async request<T>({
+    path,
+    method,
+    body,
+    hasCookies,
+  }: {
+    path: string;
+    method: Method;
+    body?: BodyHashMap;
+    hasCookies?: boolean;
+  }): Promise<T> {
+    const url = this.baseURL + path;
+    const response = await fetch(url, this.getRequestInit({ method, body, hasCookies }));
```

Comment:
`baseURL` 필드는 `URL` 타입인데 `path`는 `string` 타입이네요. 타입 안전성 등을 고려해서 아래와 같이 수정해주시면 어떨지 제안드려 봅니다 🙏 

```suggestion
    const url = new URL(path, this.baseURL);
    const response = await fetch(url.toString(), this.getRequestInit({ method, body, hasCookies }));
```